File: challenges/python/length_last_word.py
########################################################
# Length of Last Word

# Given a string s consisting of words and spaces, return the length of the last word in the string.
# A word is a maximal substring consisting of non-space characters only.

# Example 1:
# Input: s = "Hello World"
# Output: 5
# Explanation: The last word is "World" with length 5.

# Example 2:
# Input: s = "   fly me   to   the moon  "
# Output: 4
# Explanation: The last word is "moon" with length 4.

# Example 3:
# Input: s = "luffy is still joyboy"
# Output: 6
# Explanation: The last word is "joyboy" with length 6.
########################################################

#--------------------------------#
# Solution 1
# To solve this problem, we need to find the last word in the string and return its length. We can achieve this by splitting the string into words and then returning the length of the last word.

# Approach
# 1. Strip trailing whitespaces from the input string using the strip() method.
# 2. Split the string into words using the split() method.
# 3. If there are no words after stripping whitespaces, return 0.
# 4. Otherwise, return the length of the last word, which is the last element in the list of words.

# Time complexity:
# 1. Stripping trailing whitespaces takes linear time, so it's O(n) where n is the length of the input string.
# 2. Splitting the string into words also takes linear time, so it's also O(n).
# 3. The overall time complexity is O(n).
# Space complexity:
# We store the list of words, which could take up to O(n) space if all characters in the input string are non-whitespace.
#--------------------------------#
class Solution(object):
    def lengthOfLastWord(self, s):
        """
        :type s: str
        :rtype: int
        """
        words = s.strip().split()
        if not words:
          return 0
        return len(words[-1])


solution = Solution()
print(solution.lengthOfLastWord("Hello World"))
print(solution.lengthOfLastWord("   fly me   to   the moon  "))
print(solution.lengthOfLastWord("luffy is still joyboy"))
print("--------------------")



class SolutionOneLiner(object):
    def lengthOfLastWord(self, s):
        """
        :type s: str
        :rtype: int
        """
        return len((s.strip()).split(" ")[-1])

solution_one_liner = SolutionOneLiner()
print(solution_one_liner.lengthOfLastWord("Hello World"))
print(solution_one_liner.lengthOfLastWord("   fly me   to   the moon  "))
print(solution_one_liner.lengthOfLastWord("luffy is still joyboy"))
print("--------------------")

class Solution2:
    def lengthOfLastWord(self, s: str) -> int:
        lst = s.split(' ')
        for i in range(len(lst)-1,-1,-1):
            if lst[i]:
                return len(lst[i])
        return 0

solution2 = Solution2()
print(solution2.lengthOfLastWord("Hello World"))
print(solution2.lengthOfLastWord("   fly me   to   the moon  "))
print(solution2.lengthOfLastWord("luffy is still joyboy"))
print("--------------------")

# Fast Code
class Solution3(object):
    def lengthOfLastWord(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = len(s) - 1 
        length = 0

        while i >= 0 and s[i] == " ":
            i -= 1

        while i >= 0 and s[i] != " ":
            length += 1
            i -= 1

        return length

solution3 = Solution3()
print(solution3.lengthOfLastWord("Hello World"))
print(solution3.lengthOfLastWord("   fly me   to   the moon  "))
print(solution3.lengthOfLastWord("luffy is still joyboy"))
print("--------------------")

File: challenges/javascript/lengthLastWord.js
//#######################################################
// Length of Last Word

// Given a string s consisting of words and spaces, return the length of the last word in the string.
// A word is a maximal substring consisting of non-space characters only.

// Example 1:
// Input: s = "Hello World"
// Output: 5
// Explanation: The last word is "World" with length 5.

// Example 2:
// Input: s = "   fly me   to   the moon  "
// Output: 4
// Explanation: The last word is "moon" with length 4.

// Example 3:
// Input: s = "luffy is still joyboy"
// Output: 6
// Explanation: The last word is "joyboy" with length 6.
//########################################################

const lengthOfLastWord = (s) => {
  // remove leading and trailing whitespace and 
  // split a string into substrings using the specified separator and return them as an array.
  const words = s.trim().split(" ");
  const lastWord = words[words.length - 1];
  // return last word, split each letter into strings in an array and return length of array.
  return lastWord.split("").length;
};

console.log(lengthOfLastWord("Hello World"));
console.log(lengthOfLastWord("   fly me   to   the moon  "));
console.log(lengthOfLastWord("luffy is still joyboy"));
console.log("---------------------");

function lengthOfLastWord2(s) {
  s = s.trim();

  let length = 0;
  for (let i = s.length - 1; i >= 0; i--) {
    if (s[i] !== " ") {
      length++;
    } else if (length > 0) {
      break;
    }
  }

  return length;
}

console.log(lengthOfLastWord2("Hello World"));
console.log(lengthOfLastWord2("   fly me   to   the moon  "));
console.log(lengthOfLastWord2("luffy is still joyboy"));
console.log("---------------------");

/*
Time complexity: O(n)
Space complexity: O(1)
*/
var lengthOfLastWord3 = function (s) {
  let end = s.length - 1;

  while (end >= 0 && s[end] === " ") {
    end--;
  }

  let start = end;
  while (start >= 0 && s[start] !== " ") {
    start--;
  }

  return end - start;
};

console.log(lengthOfLastWord3("Hello World"));
console.log(lengthOfLastWord3("   fly me   to   the moon  "));
console.log(lengthOfLastWord3("luffy is still joyboy"));
console.log("---------------------");

/*
Initialize Variables: Prepare length and counting for tracking.
Loop Through Characters: Examine each character to determine if it's part of a word or a space.
Update Length: Adjust length based on whether you're starting a new word or continuing an existing one.
Handle Spaces: Reset counting when spaces are encountered.
Return Result: Output the length of the last word.
This approach ensures that only the length of the last word is calculated, and any trailing spaces or spaces between words do not affect the final result.
*/
var lengthOfLastWord4 = function (s) {
  let length = 0;
  let counting = false;

  for (let c of s) {
    if (c !== " ") {
      if (!counting) {
        counting = true;
        length = 1;
      } else {
        length++;
      }
    } else {
      counting = false;
    }
  }

  return length;
};

console.log(lengthOfLastWord4("Hello World"));
console.log(lengthOfLastWord4("   fly me   to   the moon  "));
console.log(lengthOfLastWord4("luffy is still joyboy"));
console.log("---------------------");
